std/math/big/int.zig

```diff
@@ -0,0 +1,2023 @@
+const std = @import("../../index.zig");
+const builtin = @import("builtin");
+const debug = std.debug;
+const math = std.math;
+const mem = std.mem;
+const Allocator = mem.Allocator;
+const ArrayList = std.ArrayList;
+
+const TypeId = builtin.TypeId;
+
+pub const Limb = usize;
+pub const DoubleLimb = @IntType(false, 2 * Limb.bit_count);
+pub const Log2Limb = math.Log2Int(Limb);
+
+comptime {
+    debug.assert(math.floorPowerOfTwo(usize, Limb.bit_count) == Limb.bit_count);
+    debug.assert(Limb.bit_count <= 64); // u128 set is unsupported
+    debug.assert(Limb.is_signed == false);
+}
+
+const wrapped_buffer_size = 512;
+
+// Converts primitive integer values onto a stack-based big integer, or passes through existing
+// Int types with no modifications. This can fail at runtime if using a very large dynamic
+// integer but it is very unlikely and is considered a user error.
+fn wrapInt(allocator: *Allocator, bn: var) *const Int {
+    const T = @typeOf(bn);
+    switch (@typeInfo(T)) {
+        TypeId.Pointer => |info| {
+            if (info.child == Int) {
+                return bn;
+            } else {
+                @compileError("cannot set Int using type " ++ @typeName(T));
+            }
+        },
+        else => {
+            var s = allocator.create(Int) catch unreachable;
+            s.* = Int{
+                .allocator = allocator,
+                .positive = false,
+                .limbs = block: {
+                    var limbs = allocator.alloc(Limb, Int.default_capacity) catch unreachable;
+                    limbs[0] = 0;
+                    break :block limbs;
+                },
+                .len = 1,
+            };
+            s.set(bn) catch unreachable;
+            return s;
+        },
+    }
+}
+
+pub const Int = struct {
```

Comment:
I have some ideas that might be worth exploring, but maybe you already considered them, and have some thoughts?

Also I don't consider any of these merge blockers.

### allocator in function params instead of Int struct

Downsides: introduces a new kind of footgun where you can use one allocator for initializing and then a different one for addition or something and that would be undefined behavior. This could be mitigated in several ways:
 * Int could include allocator as a field, but only in debug mode and assert that passed allocator matches.
 * Generally, allocators can include safety in debug mode that make note of the `*Allocator` address as metadata along side allocated data, and can assert that a free/realloc uses the same address.
 * Even more generally, the `Allocator` interface itself, could do this on top of every allocator implementation.

Another downside: less convenient from an API perspective

Benefits:
 * less memory usage
 * better cache performance?

### union for limbs to handle the case where the int is `<= @maxValue(Limb)`

This is for higher performance and to avoid using memory allocation when possible. In general we would like to use big ints when the value *could* possibly be larger than one of the builtin int sizes, but even in these cases it is likely that a `usize` could store it after all.

### try to use less memory

One example source file is `std/special/compiler_rt/udivmodti4_test.zig`. It's 65300 lines and each line has 4 bigints. At 40 bytes per big int, this comes out to 10MB. If big ints were 16 bytes instead, this comes out to 4MB, which is a pretty significant difference.

Related: #471 

Here are some ideas to conserve memory:

```zig
comptime {
    // status quo
    // 40 bytes on x86_64
    // 20 bytes on i386
    @compileLog(@sizeOf(struct {
        allocator: *u8,
        positive: bool,
        limbs: []i32,
        len: usize,
    }));
    // Use len to manage heap allocation
    // 24 bytes on x86_64
    // 12 bytes on i386
    @compileLog(@sizeOf(struct {
        limbs: packed union {
            Heap: [*]u8,
            Single: usize,
        },
        len: usize,
        positive: bool,
    }));
    // u32 for len
    // 16 bytes on x86_64
    // 12 bytes on i386
    @compileLog(@sizeOf(struct {
        limbs: packed union {
            Heap: [*]u8,
            Single: usize,
        },
        len: u32,
        positive: bool,
    }));
    // use the most significant bit of len for positive
    // 16 bytes on x86_64
    // 8 bytes on i386
    @compileLog(@sizeOf(packed struct {
        limbs: packed union {
            Heap: [*]u8,
            Single: usize,
        },
        positive: bool,
        len: @IntType(false, usize.bit_count - 1),
    }));
}
```

Reply:
I suppose this all comes down to the difference in usage of arbitrary precision integers. For mathematical/cryptographic purposes you generally don't have a large number of objects and they are typically very large so you don't need to specialize small sizes. The compiler on the other hand uses it as its default integer type and allocates many of them at once.

Let me think about this a bit and I'll get back to you. I'm still inclined to prefer the allocator field still stored with the type as is since it simplifies the usage and avoids the possible use errors as suggested. I'll see how much the implementation changes special-casing small limbs and/or changing the length representation. At the least, packing the pointer in the length is a worthwhile change.

Reply:
> Let me think about this a bit and I'll get back to you.

Sounds good.

Another argument in favor of how you did it here is that the compiler is the one which has weird memory requirements, and thus can be the one to do the tricks. For example, in the compiler the type can look something like this:

```zig
const Value = union(enum) {
    Pointer: *PointerInfo,
    Bool: bool,
    SmallSignedInt: i64,
    SmallUnsignedInt: u64,
    Int: *std.math.big.Int,
};
```

So here we've managed to keep the base `Value` type down to `@sizeOf(usize)`, at the cost of a little complexity initializing and dealing with some of the extra info that requires more bytes to store.

Reply:
Alright, I think for the moment I'll use the type:

```
struct {
    allocator: *u8,
    limbs: []u8,
    metadata: usize,
}
```

Where the top-bit of `metadata` is the sign bit, with the remaining being the length of the limbs. This is 32 bytes so a small 8 byte saving. We can drop this down to 24 bytes in the future by using a `u31/u32` for the cap and length if we want a bit more space (limiting the max `Int` size to 16gb, however).

Also note that general purpose memory allocators may not allocate certain small bin sizes (jemalloc for instance only does 16, 32, 48...) so certain byte sizes may not be as useful. This isn't always the case though, ptmalloc2 (libc) for instance looks like it does 8-byte strides so it doesn't matter there.